File: vehicle-feature-catalog/src/ts/loader.ts
import { promises as fs } from 'node:fs';
import * as path from 'node:path';
import * as yaml from 'js-yaml';

import {
  Availability,
  Feature,
  Make,
  Model,
  Trim,
  TrimFeature,
} from './entities.js';
import { LoadError } from './errors.js';

export interface RawCatalog {
  makes: Make[];
  models: Model[];
  trims: Trim[];
  features: Feature[];
  trim_features: TrimFeature[];
}

const VALID_AVAILABILITY: ReadonlySet<Availability> = new Set([
  'standard',
  'optional',
  'unavailable',
]);

export class YAMLLoader {
  static async read_all(dataDir: string): Promise<RawCatalog> {
    const stat = await safeStat(dataDir);
    if (!stat) {
      throw new LoadError(`data_dir does not exist: ${dataDir}`);
    }
    if (!stat.isDirectory()) {
      throw new LoadError(`data_dir is not a directory: ${dataDir}`);
    }

    const makes = await this.loadMakes(path.join(dataDir, 'makes'));
    const models = await this.loadModels(path.join(dataDir, 'models'));
    const trims = await this.loadTrims(path.join(dataDir, 'trims'));
    const features = await this.loadFeatures(path.join(dataDir, 'features'));
    const trim_features = await this.loadMatrix(path.join(dataDir, 'matrix'));

    return { makes, models, trims, features, trim_features };
  }

  private static async loadMakes(dir: string): Promise<Make[]> {
    const out: Make[] = [];
    for (const filePath of await listYamlFiles(dir)) {
      const data = await readYaml(filePath);
      requireKeys(filePath, data, ['id', 'name', 'country']);
      out.push({
        id: String(data['id']),
        name: String(data['name']),
        country: String(data['country']),
      });
    }
    return out;
  }

  private static async loadModels(dir: string): Promise<Model[]> {
    const out: Model[] = [];
    for (const filePath of await listYamlFiles(dir)) {
      const data = await readYaml(filePath);
      requireKeys(filePath, data, ['id', 'make_id', 'name', 'year']);
      const body = data['body_style'];
      out.push({
        id: String(data['id']),
        make_id: String(data['make_id']),
        name: String(data['name']),
        year: Number(data['year']),
        body_style: body == null ? null : String(body),
      });
    }
    return out;
  }

  private static async loadTrims(dir: string): Promise<Trim[]> {
    const out: Trim[] = [];
    for (const filePath of await listYamlFiles(dir)) {
      const data = await readYaml(filePath);
      requireKeys(filePath, data, ['id', 'model_id', 'name']);
      const raw = data['msrp_range'];
      let msrp_range: [number, number] | null;
      if (raw == null) {
        msrp_range = null;
      } else if (Array.isArray(raw) && raw.length === 2) {
        msrp_range = [Number(raw[0]), Number(raw[1])];
      } else {
        throw new LoadError(
          `${filePath}: msrp_range must be a 2-element list, got ${JSON.stringify(raw)}`,
        );
      }
      out.push({
        id: String(data['id']),
        model_id: String(data['model_id']),
        name: String(data['name']),
        msrp_range,
      });
    }
    return out;
  }

  private static async loadFeatures(dir: string): Promise<Feature[]> {
    const out: Feature[] = [];
    for (const filePath of await listYamlFiles(dir)) {
      const data = await readYaml(filePath);
      requireKeys(filePath, data, ['id', 'display_name', 'category', 'brand_scope']);
      const cueRaw = data['cue_phrases'] ?? [];
      const synRaw = data['synonyms'] ?? [];
      if (!Array.isArray(cueRaw)) {
        throw new LoadError(`${filePath}: cue_phrases must be a list`);
      }
      if (!Array.isArray(synRaw)) {
        throw new LoadError(`${filePath}: synonyms must be a list`);
      }
      out.push({
        id: String(data['id']),
        display_name: String(data['display_name']),
        category: String(data['category']),
        brand_scope: String(data['brand_scope']),
        cue_phrases: cueRaw.map((p) => String(p)),
        synonyms: synRaw.map((s) => String(s)),
      });
    }
    return out;
  }

  private static async loadMatrix(dir: string): Promise<TrimFeature[]> {
    const out: TrimFeature[] = [];
    const stat = await safeStat(dir);
    if (!stat) return out;
    for (const filePath of await listYamlFiles(dir)) {
      const data = await readYaml(filePath);
      const entries = data['entries'];
      if (entries == null) {
        throw new LoadError(`${filePath}: matrix file missing top-level 'entries' key`);
      }
      if (!Array.isArray(entries)) {
        throw new LoadError(`${filePath}: 'entries' must be a list`);
      }
      for (const entry of entries) {
        if (typeof entry !== 'object' || entry === null) {
          throw new LoadError(`${filePath}: matrix entry must be a mapping`);
        }
        const entryRec = entry as Record<string, unknown>;
        const trimId = entryRec['trim_id'];
        const features = entryRec['features'] ?? [];
        if (!trimId) {
          throw new LoadError(`${filePath}: matrix entry missing trim_id`);
        }
        if (!Array.isArray(features)) {
          throw new LoadError(`${filePath}: features in entry must be a list`);
        }
        for (const cell of features) {
          if (typeof cell !== 'object' || cell === null) {
            throw new LoadError(
              `${filePath}: feature cell must be a mapping, got ${JSON.stringify(cell)}`,
            );
          }
          const cellRec = cell as Record<string, unknown>;
          const featureId = cellRec['feature_id'];
          const availability = cellRec['availability'];
          if (!featureId) {
            throw new LoadError(
              `${filePath}: feature cell missing feature_id under trim ${String(trimId)}`,
            );
          }
          if (
            typeof availability !== 'string' ||
            !VALID_AVAILABILITY.has(availability as Availability)
          ) {
            throw new LoadError(
              `${filePath}: invalid availability ${JSON.stringify(availability)} for ` +
                `${String(trimId)} / ${String(featureId)}`,
            );
          }
          out.push({
            trim_id: String(trimId),
            feature_id: String(featureId),
            availability: availability as Availability,
          });
        }
      }
    }
    return out;
  }
}

async function safeStat(target: string): Promise<import('fs').Stats | null> {
  try {
    return await fs.stat(target);
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') return null;
    throw err;
  }
}

async function listYamlFiles(root: string): Promise<string[]> {
  const stat = await safeStat(root);
  if (!stat) return [];
  const out: string[] = [];
  await walk(root, out);
  out.sort();
  return out;
}

async function walk(dir: string, out: string[]): Promise<void> {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      await walk(full, out);
    } else if (entry.isFile() && entry.name.endsWith('.yaml')) {
      out.push(full);
    }
  }
}

async function readYaml(filePath: string): Promise<Record<string, unknown>> {
  let text: string;
  try {
    text = await fs.readFile(filePath, 'utf-8');
  } catch (err) {
    throw new LoadError(`Failed to read ${filePath}: ${(err as Error).message}`);
  }
  let data: unknown;
  try {
    data = yaml.load(text);
  } catch (err) {
    throw new LoadError(`YAML parse error in ${filePath}: ${(err as Error).message}`);
  }
  if (data == null) {
    throw new LoadError(`${filePath}: file is empty`);
  }
  if (typeof data !== 'object' || Array.isArray(data)) {
    throw new LoadError(`${filePath}: top-level YAML must be a mapping`);
  }
  return data as Record<string, unknown>;
}

function requireKeys(
  filePath: string,
  data: Record<string, unknown>,
  keys: string[],
): void {
  const missing = keys.filter((k) => !(k in data));
  if (missing.length > 0) {
    throw new LoadError(`${filePath}: missing required key(s): ${missing.join(', ')}`);
  }
}

File: vehicle-feature-catalog/src/python/vehicle_feature_catalog/loader.py
"""YAML loader.

Walks `data_dir` and produces lists of parsed entities. Pure I/O + parsing.
Validation, indexing, and de-duplication are downstream concerns.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

from .entities import Feature, Make, Model, Trim, TrimFeature
from .errors import LoadError


@dataclass
class RawCatalog:
    makes: list[Make]
    models: list[Model]
    trims: list[Trim]
    features: list[Feature]
    trim_features: list[TrimFeature]


class YAMLLoader:
    @classmethod
    def read_all(cls, data_dir: Path) -> RawCatalog:
        if not data_dir.exists():
            raise LoadError(f"data_dir does not exist: {data_dir}")
        if not data_dir.is_dir():
            raise LoadError(f"data_dir is not a directory: {data_dir}")

        makes = cls._load_makes(data_dir / "makes")
        models = cls._load_models(data_dir / "models")
        trims = cls._load_trims(data_dir / "trims")
        features = cls._load_features(data_dir / "features")
        trim_features = cls._load_matrix(data_dir / "matrix")

        return RawCatalog(
            makes=makes,
            models=models,
            trims=trims,
            features=features,
            trim_features=trim_features,
        )

    @classmethod
    def _load_makes(cls, makes_dir: Path) -> list[Make]:
        out: list[Make] = []
        for path in cls._iter_yaml_files(makes_dir):
            data = cls._read_yaml(path)
            cls._require_keys(path, data, ["id", "name", "country"])
            out.append(
                Make(
                    id=str(data["id"]),
                    name=str(data["name"]),
                    country=str(data["country"]),
                )
            )
        return out

    @classmethod
    def _load_models(cls, models_dir: Path) -> list[Model]:
        out: list[Model] = []
        for path in cls._iter_yaml_files(models_dir):
            data = cls._read_yaml(path)
            cls._require_keys(path, data, ["id", "make_id", "name", "year"])
            body_style = data.get("body_style")
            out.append(
                Model(
                    id=str(data["id"]),
                    make_id=str(data["make_id"]),
                    name=str(data["name"]),
                    year=int(data["year"]),
                    body_style=str(body_style) if body_style is not None else None,
                )
            )
        return out

    @classmethod
    def _load_trims(cls, trims_dir: Path) -> list[Trim]:
        out: list[Trim] = []
        for path in cls._iter_yaml_files(trims_dir):
            data = cls._read_yaml(path)
            cls._require_keys(path, data, ["id", "model_id", "name"])
            msrp_range_raw = data.get("msrp_range")
            msrp_range: tuple[int, int] | None
            if msrp_range_raw is None:
                msrp_range = None
            else:
                if not (isinstance(msrp_range_raw, (list, tuple)) and len(msrp_range_raw) == 2):
                    raise LoadError(
                        f"{path}: msrp_range must be a 2-element list, got {msrp_range_raw!r}"
                    )
                msrp_range = (int(msrp_range_raw[0]), int(msrp_range_raw[1]))
            out.append(
                Trim(
                    id=str(data["id"]),
                    model_id=str(data["model_id"]),
                    name=str(data["name"]),
                    msrp_range=msrp_range,
                )
            )
        return out

    @classmethod
    def _load_features(cls, features_dir: Path) -> list[Feature]:
        out: list[Feature] = []
        for path in cls._iter_yaml_files(features_dir):
            data = cls._read_yaml(path)
            cls._require_keys(
                path,
                data,
                ["id", "display_name", "category", "brand_scope"],
            )
            cue_phrases = data.get("cue_phrases") or []
            synonyms = data.get("synonyms") or []
            if not isinstance(cue_phrases, list):
                raise LoadError(f"{path}: cue_phrases must be a list")
            if not isinstance(synonyms, list):
                raise LoadError(f"{path}: synonyms must be a list")
            out.append(
                Feature(
                    id=str(data["id"]),
                    display_name=str(data["display_name"]),
                    category=str(data["category"]),
                    brand_scope=str(data["brand_scope"]),
                    cue_phrases=[str(p) for p in cue_phrases],
                    synonyms=[str(s) for s in synonyms],
                )
            )
        return out

    @classmethod
    def _load_matrix(cls, matrix_dir: Path) -> list[TrimFeature]:
        out: list[TrimFeature] = []
        if not matrix_dir.exists():
            return out
        for path in cls._iter_yaml_files(matrix_dir):
            data = cls._read_yaml(path)
            entries = data.get("entries")
            if entries is None:
                raise LoadError(f"{path}: matrix file missing top-level 'entries' key")
            if not isinstance(entries, list):
                raise LoadError(f"{path}: 'entries' must be a list")
            for entry in entries:
                if not isinstance(entry, dict):
                    raise LoadError(f"{path}: matrix entry must be a mapping")
                trim_id = entry.get("trim_id")
                features = entry.get("features") or []
                if not trim_id:
                    raise LoadError(f"{path}: matrix entry missing trim_id")
                if not isinstance(features, list):
                    raise LoadError(f"{path}: features in entry must be a list")
                for cell in features:
                    if not isinstance(cell, dict):
                        raise LoadError(
                            f"{path}: feature cell must be a mapping, got {cell!r}"
                        )
                    feature_id = cell.get("feature_id")
                    availability = cell.get("availability")
                    if not feature_id:
                        raise LoadError(
                            f"{path}: feature cell missing feature_id under trim {trim_id}"
                        )
                    if availability not in ("standard", "optional", "unavailable"):
                        raise LoadError(
                            f"{path}: invalid availability {availability!r} for "
                            f"{trim_id} / {feature_id}"
                        )
                    out.append(
                        TrimFeature(
                            trim_id=str(trim_id),
                            feature_id=str(feature_id),
                            availability=availability,
                        )
                    )
        return out

    @staticmethod
    def _iter_yaml_files(root: Path) -> list[Path]:
        if not root.exists():
            return []
        return sorted(p for p in root.rglob("*.yaml") if p.is_file())

    @staticmethod
    def _read_yaml(path: Path) -> dict[str, Any]:
        try:
            with path.open("r", encoding="utf-8") as fh:
                data = yaml.safe_load(fh)
        except yaml.YAMLError as exc:
            raise LoadError(f"YAML parse error in {path}: {exc}") from exc
        except OSError as exc:
            raise LoadError(f"Failed to read {path}: {exc}") from exc
        if data is None:
            raise LoadError(f"{path}: file is empty")
        if not isinstance(data, dict):
            raise LoadError(f"{path}: top-level YAML must be a mapping, got {type(data).__name__}")
        return data

    @staticmethod
    def _require_keys(path: Path, data: dict[str, Any], keys: list[str]) -> None:
        missing = [k for k in keys if k not in data]
        if missing:
            raise LoadError(f"{path}: missing required key(s): {', '.join(missing)}")
